### DISCUSSION TITLE
How can I change font of verbatim content?

### DISCUSSION THREAD

**Discussion by nawordar** [Omikhleia/resilient.sile#55] [Q&A] [answered]
I am creating a document using the resilient book class with djot as a markup language. Is there a way to change the font of verbatim content? I see that the `markdown.sile` package calls SILE's `code` and `verbatim` commands, but I don't know how I could redefine them when using the `.silm` file.

**Comment by Omikhleia** [comment, marked as answer]
Greetings,

Good question. _Eventually_ in the "resilient" scheme, this ought to be done with styles, possibly with inheritance, to cover both inline code and code blocks -- Currently they just rely on SILE's `code` and `verbatim` directly, respectively -- a quick and dirty choice (They both pick the Hack font directly and independently, in SILE's core implementation)

To be honest I did not work much on these, as my main focus currently is in on novels, which seldom need code (inline or verbatim) ;)

(Code blocks also ought to have a better solution for syntax highlighting, currently only supported for Lua with a naive approach and a hard-coded theme, but that's a broader topic). 

A workaround for your question, currently, is to add a regular `.sil` file at the beginning of your master document (`.silm`), and to override the `verbatim:font` hook as you would do with regular SIL (`\define[command=verbatim:font]{\font[family=...]}`) -- But it's only a workaround as it would likely break when we move to proper styles eventually.

I'll change this question into an enhancement issue and try to see if I can address it properly with styles in a later version. Not sure it can fit in the next version (planned around October or November), but it should definitively be tackled at one point.

**Reply by nawordar** [reply]
I didn't manage to post this update before your answer, though your solution would be definitely simpler :):

I managed to create a resilient-compatible class and even add my own entry in the style file. This entry only works with inline code, though, not with code blocks:

```lua
local book = require 'classes.resilient.book'

local class = pl.class(book)
class._name = 'resilient-notes'

function class:_init(options)
  book._init(self, options)

  self:registerStyles()
  self:applyStyles()
end

function class:registerStyles()
  book.registerStyles(self)

  self:registerStyle('verbatim', {}, {
    font = { family = 'Hack', size = '1.4ex' },
  })
end

function class:applyStyles()
  self:registerCommand('code', function(_, content)
    SILE.call('style:apply', { name = 'verbatim' }, content)
  end)

  self:registerCommand('verbatim:font', function(_, content)
    -- Doesn't work
    -- SILE.call('style:apply', { name = 'verbatim' }, content)
    SILE.call('font', { family = 'JetBrains Mono', size = '0.8em' })
  end)
end

return class
```

**Reply by nawordar** [reply]
By the way, do you plan on moving the default font settings to styles too? That is, the ones that are currently set through the `font` key in the `silm` file.

**Reply by Omikhleia** [reply]
> I managed to create a resilient-compatible class and even add my own entry in the style file

Wow, pretty nice workaround solution too, congrats with it.
(as for the "Doesn't work" comment in your code, yes, that's part of the issue, the verbatim:font hook sets a global font with a temporary environment -- all these low-levels hooks work differently...)

> By the way, do you plan on moving the default font settings to styles too? That is, the ones that are currently set through the font key in the silm file.

I would have wished initially (e.g. to have some global "document" style for such things)... The main problem I had, though, is font fallbacks, which aren't scoped yet (https://github.com/sile-typesetter/sile/issues/1543)...
